scalding-core/src/test/scala/com/twitter/scalding/typed/OptimizationRulesTest.scala

```diff
@@ -644,4 +661,50 @@ class OptimizationRulesTest extends FunSuite with PropertyChecks {
 
     optimizedSteps(OptimizationRules.standardMapReduceRules, 1)(pipe)
   }
+
+  test("dediamond map only diamonds") {
```

Comment:
could we have tests to cover all or most of cases (Filter, FlatMap, MapValues, etc)?

Reply:
yep. Happy to add more tests. That's a good call. Thank you.

Reply:
This was a good suggestion. There are some ordering issues that come into play that this detected. For instance, you don't want to push mappers up into reducers or filterKey above a group until after we have done the de-diamond, else it can make a diamond look not like a diamond by making the reduce look different.

Reply:
nice!